### PR TITLE
BUG: fixes exception in numpy.genfromtxt, see #10780

### LINKED ISSUE #10780: In genfromtxt(), comments=None raises error with names=True
Sorry in advance if this is a duplicate. As of numpy 1.12.1,  an error is raised by genfromtxt() when using `names=True` with `comments=None`. Tested with:
* Python 2.7.14 |Anaconda custom (64-bit)| (default, Oct 16 2017, 17:29:19) 
* numpy 1.12.1
* Linux rsiverd-linux.lco.gtn 3.10.0-693.5.2.el7.x86_64 #1 SMP Fri Oct 20 20:32:50 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

...............................................................................................................

**Instructions to reproduce** (using attached sample file 'asdf.txt'):
```
data = np.genfromtxt('asdf.txt', dtype=None, comments=None, delimiter='|', names=True)
```
[asdf.txt](https://github.com/numpy/numpy/files/1835574/asdf.txt)
...............................................................................................................
**Expected result**: 
* Successful loading of the data file (no error).

...............................................................................................................
**Actual result**:
```
TypeError                                 Traceback (most recent call last)
<ipython-input-1-1d47ac60e22b> in <module>()
----> 1 data = np.genfromtxt('asdf.txt', dtype=None, comments=None, delimiter='|', names=True)

/home/rsiverd/anaconda2/lib/python2.7/site-packages/numpy/lib/npyio.pyc in genfromtxt(fname, dtype, comments, delimiter, skip_header, skip_footer, converters, missing_values, filling_values, usecols, names, excludelist, deletechars, replace_space, autostrip, case_sensitive, defaultfmt, unpack, usemask, loose, invalid_raise, max_rows)
   1536             first_line = next(fhd)
   1537             if names is True:
-> 1538                 if comments in first_line:
   1539                     first_line = (
   1540                         asbytes('').join(first_line.split(comments)[1:]))

TypeError: 'in <string>' requires string as left operand, not NoneType
```

...............................................................................................................
**Suggested remedy:**

It looks like the check for the first valid values does not check whether `comments is None` before looking for the comment string in the first line, raising this error. I suspect this is easily fixed by replacing
```
        while not first_values:
            first_line = next(fhd)
            if names is True:
                if comments in first_line:
                    first_line = (
                        asbytes('').join(first_line.split(comments)[1:]))
            first_values = split_line(first_line)
```
with
```
        while not first_values:
            first_line = next(fhd)
            if (names is True) and (comments is not None):
                if comments in first_line:
                    first_line = (
                        asbytes('').join(first_line.split(comments)[1:]))
            first_values = split_line(first_line)
```

### PR DESCRIPTION
**Pull request by raunaks13**
Handles case when names is set to True and comments is set to None by the user.

Closes #10780

### CODE REVIEW

**Review by eric-wieser — CHANGES_REQUESTED** [review]
Needs a test in `numpy\lib\tests\test_io.py`

**Comment by eric-wieser** [comment]
Your fix looks correct, but the code it's fixing still looks super broken...

Why do we check `comment in line` in one place, yet `value in comment` in another? One is treating `comment` as a string, the other as a collection of single characters!

Not something you need to deal with in this PR, but I

**Comment by raunaks13** [comment]
@eric-wieser I'm a bit new to this so would this function in test_io.py suffice?:
```python
    def test_names_and_comments_none(self):
        #tests case when names is true but comments is none
        data = TextIO('col1 col2\n 1 2\n 3 4')
        test = np.genfromtxt(data, dtype=(int, int), comments=None, names=True)
        control = np.array([[1, 2], [3, 4]], dtype=[('col1', int), ('col2', int)])
        assert_equal(test, control)

File: numpy/lib/tests/test_io.py [hunk 1]
```diff
@@ -1272,6 +1272,15 @@ def test_commented_header(self):
             assert_(w[0].category is np.VisibleDeprecationWarning)
         assert_equal(test, ctrl)
 
+    def test_names_and_comments_none(self):
+        #tests case when names is true but comments is None
+        data = TextIO('col1 col2\n 1 2\n 3 4')
+        with warnings.catch_warnings(record=True) as w:
+            warnings.filterwarnings('always', '', np.VisibleDeprecationWarning)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_io.py]: What's going on here?

Comment (raunaks13) [reply numpy/lib/tests/test_io.py]: To catch warnings...if it's not necessary I'll remove those 2 lines

Comment (eric-wieser) [reply numpy/lib/tests/test_io.py]: What warnings are you producing?

Comment (raunaks13) [reply numpy/lib/tests/test_io.py]: This code is present in npyio.py

File: numpy/lib/tests/test_io.py [hunk 2]
```diff
@@ -1272,6 +1272,13 @@ def test_commented_header(self):
             assert_(w[0].category is np.VisibleDeprecationWarning)
         assert_equal(test, ctrl)
 
+    def test_names_and_comments_none(self):
+        #tests case when names is true but comments is None
```
Comment (eric-wieser) [inline numpy/lib/tests/test_io.py]: One final comment - it's nice to reference the github issue with `gh-10780` in test comments

**Review by eric-wieser — DISMISSED** [review]

**Comment by raunaks13** [comment]
@eric-wieser Any other changes necessary? Or could we merge this?

**Comment by raunaks13** [comment]
@eric-wieser Previously in issue #329 it was proposed to add support for `comments=None` since there is a chance of `None` popping up in large datasets. PR #337 added support and was merged as well.

Of course this bug remained so it seems logical we should fix it now.

**Comment by eric-wieser** [comment]
Thanks for the link, that's exactly the kind of precedent I was looking for.

Can I request that you add a remark to the docstring for the  `comments`  argument? Since it looks like (partial) support was added for this in 1.7, maybe:

```
The character used to indicate the start of a comment. All the characters occurring on a line after a comment are discarded.

.. versionadded:: 1.7.0

When None, no lines are treated as comments.
```

File: doc/source/user/basics.io.genfromtxt.rst [hunk 3]
```diff
@@ -127,19 +127,22 @@ marker(s) is simply ignored::
...
+   [[ 1.  2.
     [ 3.  4.]
     [ 5.  6.]
     [ 7.  8.]
     [ 9.  0.]]
 
+.. versionadded:: 1.7.0
+
+    When ``comments`` is set to ``None``, no lines are treated as comments.
+
```
Comment (eric-wieser) [inline doc/source/user/basics.io.genfromtxt.rst]: This ought to go under the other description of `comments`

File: doc/source/user/basics.io.genfromtxt.rst [hunk 4]
```diff
@@ -127,19 +127,22 @@ marker(s) is simply ignored::
    ... 9, 0
    ... """
    >>> np.genfromtxt(BytesIO(data), comments="#", delimiter=",")
-   [[ 1.  2.]
+   [[ 1.  2.
```
Comment (eric-wieser) [inline doc/source/user/basics.io.genfromtxt.rst]: Looks like this happened by mistake

**Comment by eric-wieser** [comment]
Thanks for the persistence, @raunaks13!

**Pull request merged**